hangupsbot/commands/__init__.py

```diff
@@ -18,6 +21,182 @@ def __init__(self):
 
         self.command_tagsets = {}
 
+        self.preprocessors = { "inbuilt": { r"(?<!@)@[^@]\w+[^@]$": self.one_chat_id,
+                                            r"(?<!#)#[^#]\w+[^#]$": self.one_conv_id }}
+
+    def one_chat_id(self, token, internal_context):
+        user_memory = self.bot.get_memory_option("user_data")
+        if internal_context:
+            chat_ids = list(self.bot.conversations.catalog[internal_context.conv_id]["participants"])
+        else:
+            chat_ids = list(user_memory.keys())
+
+        fragment = token[1:]
+        matched_users = {}
+        for chat_id in chat_ids:
+            user_data = user_memory[chat_id]
+            if "_hangups" in user_data:
+                if "nickname" in user_data:
+                    nickname_lower =  user_data["nickname"].lower()
+                else:
+                    nickname_lower = ""
+                fullname_lower = user_data["_hangups"]["full_name"].lower()
+
+                if fragment == nickname_lower:
+                    matched_users[chat_id] = chat_id
+                    break
+
+                elif( fragment in fullname_lower or
+                        fragment in fullname_lower.replace(" ", "") ):
+                    matched_users[chat_id] = chat_id
+
+        if len(matched_users) == 1:
+            return list(matched_users)[0]
+        elif len(matched_users) == 0:
+            if internal_context:
+                return self.one_chat_id(token, False)
+            else:
+                raise ValueError("{} returned no users".format(token))
+        else:
+            raise ValueError("{} returned more than one user".format(token))
+
+    def one_conv_id(self, token, internal_context):
+        text = token[1:]
+
+        if text == "here":
+            # current conversation id
+            return internal_context.conv_id
+
```

Comment:
the preprocessors should not affect previous usage of "here", since it expects the `#` prefix. previous implementations can continue to use "here" safely, although time will tell whether this is really the case or there is some obscure bug somewhere we haven't found yet